rfcs/2025-05-05-client-certificates.md

```diff
@@ -0,0 +1,129 @@
+
+Each transport block can contain one or several protocol transmissions.
+
+Each transmission has this structure:
+
+```abnf
+transmission = authenticator authorized
+; authenticator - Ed25519 signature for recipients or X25519 authenticator for senders, to provide repudiation.
+; authenticator authorizes the rest of the transmission.
+authorized = sessId corrId entityId command.
+; sessId is tls-unique channel binding, its presense in the transmission prevents replay attacks.
+```
+
+The proposed change would replace authenticator eith with an array of authenticators or with exactly one or two authenticators, where the first one will remain resource-level authorization (queue key), and the second - optional one - would be client authorization with client key.
+
+```abnf
+authenticator = queue_authenticator ("0" / "1" client_authenticator)
+; "0" and "1" characters (digit characters, not x00 or x01) are conventionally used for Maybe types in the protocol.
+```
+
+All queues created with client key will have to be double-authorized with both the queue key and the client key - both the client and the server would have to maintain this knowledge, whether the queue is associated with the client or not.
+
+Asymmetric retries have to be considered - the first request creating this association may succeed on the server and timeout on the client.
+
+## Subscription
+
+There are two options:
+
+- automatically subscribe client on connection in case certificate is introduced. This avoids additional command for the client, but it may work for considered ephemeral session associations with the usual clients (see below) for iOS NSE clients.
+- send a separate command authorized with the client key passed in handshake.
+
+Command response pattern seems better for ephemeral session associations, and more consistent with the protocol design.
+
+The command and response:
+
+```haskell
+CSUB :: Command Recipient -- to enable all client subscriptions, empty entity ID in the transmission, signed by client key - it must be the same as was used in handover subscription signature.
+CSQS :: Word32 -> BrokerMsg -- response from the server, includes the number of subscribed queues
+```
+
+## Ephemeral client-session association
+
+This was considered to reduce costs for the usual clients to re-subscribe. Currently it's a big problem, because of groups, and with transition to chat relays it won't be.
+
+For some very busy end-user clients it may help.
+
+Given that server has access to an ephemeral association between recipient client session and queues anyway (even with clients connecting via Tor, unless per-connection transport isolation is used), introducing `sessionPubKey` to allow resubscription to the previously subscribed queues may reduce the traffic. This won't change threat model as the server would only keep this association in memory, and not persist it. Clients on another hand may safely persist this association for fast resubscription on client restarts.
```

Comment:
server keeping it in memory seems more like a promise and not hard thread model

Reply:
"threat". Yes. But the point is that ephemeral session certificates don't change it in any way, as the server can track queues that clients subscribe to.

As for "hardness" of promise, TEEs can make it harder.